--- a/main.c
+++ b/main.c
@@ -124,6 +124,7 @@
   char *fontface;
   FcPattern *fc_pattern;
   FcBool scalable;
+  FcResult res;
   
   fc_pattern = FcNameParse((unsigned char *)list);
 
@@ -140,7 +141,7 @@
   // fill in missing defaults
   FcDefaultSubstitute(fc_pattern);
   // find a font face on our list if possible
-  fc_pattern = FcFontMatch(0, fc_pattern, 0);
+  fc_pattern = FcFontMatch(0, fc_pattern, &res);
 
   if(!fc_pattern){
     fprintf(stderr,_("\nUnable to find any suitable %s fonts!\n"
--- a/Makefile
+++ b/Makefile
@@ -89,9 +90,10 @@
 include $(SRC:.c=.d)
 endif
 
-target:  $(OBJ) 
+target: $(TARGET)
+$(TARGET): $(OBJ)
 	./touch-version
-	$(LD) $(OBJ) $(CFLAGS) -o $(TARGET) $(LIBS) $(LDF)
+	$(LD) $(OBJ) $(LDFLAGS) -o $(TARGET) $(LIBS) $(LDF) -lm
 
 install: install-local install-recursive
 
--- a/nls.h
+++ b/nls.h
@@ -1,5 +1,4 @@
 #ifdef ENABLE_NLS
-#include "gettext.h"
 #include <locale.h>
 #include <libintl.h>
 
