--- a/src/requests/certs.py
+++ b/src/requests/certs.py
@@ -11,7 +11,15 @@
 environment, you can change the definition of where() to return a separately
 packaged CA bundle.
 """
-from certifi import where
+
+import os
+
+if "_PIP_STANDALONE_CERT" not in os.environ:
+    def where():
+        return "/etc/pki/tls/certs/ca-bundle.crt"
+else:
+    def where():
+        return os.environ["_PIP_STANDALONE_CERT"]
 
 if __name__ == "__main__":
     print(where())
--- a/setup.cfg
+++ b/setup.cfg
@@ -4,7 +4,6 @@
 	socks
 	use_chardet_on_py3
 requires-dist = 
-	certifi>=2017.4.17
 	charset_normalizer>=2,<4
 	idna>=2.5,<4
 	urllib3>=1.21.1,<1.27
--- a/setup.py
+++ b/setup.py
@@ -62,7 +62,6 @@
     "charset_normalizer>=2,<4",
     "idna>=2.5,<4",
     "urllib3>=1.21.1,<3",
-    "certifi>=2017.4.17",
 ]
 test_requirements = [
     "pytest-httpbin==0.0.7",
